File: tumbuhin_flutter/lib/shared/services/personal_finance_service.dart
import '../../../core/api/api_client.dart';
import '../models/personal_budget.dart';
import '../models/financial_goal.dart';
import '../models/recurring_transaction.dart';

class PersonalFinanceService {
  final ApiClient _apiClient;
  PersonalFinanceService(this._apiClient);

  Future<Map<String, dynamic>> getSummary({int? month, int? year}) async {
    final params = <String, String>{};
    if (month != null) params['month'] = month.toString();
    if (year != null) params['year'] = year.toString();
    final response = await _apiClient.dio.get(
      '/api/v1/personal/summary',
      queryParameters: params,
    );
    return response.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> getNetWorth() async {
    final response = await _apiClient.dio.get('/api/v1/personal/net-worth');
    return response.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> recordIncome(Map<String, dynamic> data) async {
    final response = await _apiClient.dio.post(
      '/api/v1/personal/income',
      data: data,
    );
    return response.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> recordExpense(Map<String, dynamic> data) async {
    final response = await _apiClient.dio.post(
      '/api/v1/personal/expense',
      data: data,
    );
    return response.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> transfer(Map<String, dynamic> data) async {
    final response = await _apiClient.dio.post(
      '/api/v1/personal/transfer',
      data: data,
    );
    return response.data as Map<String, dynamic>;
  }

  Future<List<PersonalBudget>> getBudgets({int? month, int? year}) async {
    final params = <String, String>{};
    if (month != null) params['month'] = month.toString();
    if (year != null) params['year'] = year.toString();
    final response = await _apiClient.dio.get(
      '/api/v1/personal/budgets',
      queryParameters: params,
    );
    return (response.data as List)
        .map((e) => PersonalBudget.fromJson(e))
        .toList();
  }

  Future<Map<String, dynamic>> upsertBudget(Map<String, dynamic> data) async {
    final response = await _apiClient.dio.post(
      '/api/v1/personal/budgets',
      data: data,
    );
    return response.data as Map<String, dynamic>;
  }

  Future<List<FinancialGoal>> getGoals() async {
    final response = await _apiClient.dio.get('/api/v1/personal/goals');
    return (response.data as List)
        .map((e) => FinancialGoal.fromJson(e))
        .toList();
  }

  Future<Map<String, dynamic>> createGoal(Map<String, dynamic> data) async {
    final response = await _apiClient.dio.post(
      '/api/v1/personal/goals',
      data: data,
    );
    return response.data as Map<String, dynamic>;
  }

  Future<FinancialGoal> getGoalDetail(String id) async {
    final response = await _apiClient.dio.get('/api/v1/personal/goals/$id');
    return FinancialGoal.fromJson(response.data);
  }

  Future<Map<String, dynamic>> updateGoalProgress(
    String id,
    Map<String, dynamic> data,
  ) async {
    final response = await _apiClient.dio.patch(
      '/api/v1/personal/goals/$id/progress',
      data: data,
    );
    return response.data as Map<String, dynamic>;
  }

  Future<void> cancelGoal(String id) async {
    await _apiClient.dio.patch('/api/v1/personal/goals/$id/cancel');
  }

  Future<List<RecurringTransaction>> getRecurring() async {
    final response = await _apiClient.dio.get('/api/v1/personal/recurring');
    return (response.data as List)
        .map((e) => RecurringTransaction.fromJson(e))
        .toList();
  }

  Future<Map<String, dynamic>> createRecurring(
    Map<String, dynamic> data,
  ) async {
    final response = await _apiClient.dio.post(
      '/api/v1/personal/recurring',
      data: data,
    );
    return response.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> updateRecurring(
    String id,
    Map<String, dynamic> data,
  ) async {
    final response = await _apiClient.dio.patch(
      '/api/v1/personal/recurring/$id',
      data: data,
    );
    return response.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> triggerRecurring(String id) async {
    final response = await _apiClient.dio.patch(
      '/api/v1/personal/recurring/$id/trigger',
    );
    return response.data as Map<String, dynamic>;
  }

  Future<void> deactivateRecurring(String id) async {
    await _apiClient.dio.delete('/api/v1/personal/recurring/$id');
  }
}

File: web/src/lib/api/personalFinanceService.ts
import { apiGet, apiPost, apiPatch, apiDelete } from './client';
import type { PersonalMonthlySummary, PersonalBudget, FinancialGoal, RecurringTransaction, NetWorth } from '@/types';

export const personalFinanceService = {
  getSummary: (month?: number, year?: number) => {
    const params: Record<string, string> = {};
    if (month) params.month = String(month);
    if (year) params.year = String(year);
    return apiGet<PersonalMonthlySummary>('/api/v1/personal/summary', params);
  },

  getNetWorth: () => apiGet<NetWorth>('/api/v1/personal/net-worth'),

  recordIncome: (data: { amount: number; income_account_id: string; destination_account_id: string; date?: string; notes?: string }) =>
    apiPost<any>('/api/v1/personal/income', data),

  recordExpense: (data: { amount: number; expense_account_id: string; payment_account_id: string; date?: string; notes?: string }) =>
    apiPost<any>('/api/v1/personal/expense', data),

  transfer: (data: { amount: number; from_account_id: string; to_account_id: string; notes?: string }) =>
    apiPost<any>('/api/v1/personal/transfer', data),

  getBudgets: (month?: number, year?: number) => {
    const params: Record<string, string> = {};
    if (month) params.month = String(month);
    if (year) params.year = String(year);
    return apiGet<PersonalBudget[]>('/api/v1/personal/budgets', params);
  },

  upsertBudget: (data: { account_id: string; month: number; year: number; budget_amount: number }) =>
    apiPost<PersonalBudget>('/api/v1/personal/budgets', data),

  getGoals: () => apiGet<FinancialGoal[]>('/api/v1/personal/goals'),

  createGoal: (data: { name: string; goal_type: string; target_amount: number; target_date?: string; linked_account_id?: string; notes?: string }) =>
    apiPost<FinancialGoal>('/api/v1/personal/goals', data),

  getGoalDetail: (id: string) => apiGet<FinancialGoal>(`/api/v1/personal/goals/${id}`),

  updateGoalProgress: (id: string, data: { amount: number; source_account_id?: string; notes?: string }) =>
    apiPatch<any>(`/api/v1/personal/goals/${id}/progress`, data),

  cancelGoal: (id: string) => apiPatch<any>(`/api/v1/personal/goals/${id}/cancel`),

  getRecurring: () => apiGet<RecurringTransaction[]>('/api/v1/personal/recurring'),

  createRecurring: (data: { name: string; amount: number; direction: string; debit_account_id: string; credit_account_id: string; frequency: string; day_of_period?: number; next_due_date: string }) =>
    apiPost<RecurringTransaction>('/api/v1/personal/recurring', data),

  updateRecurring: (id: string, data: Partial<RecurringTransaction>) =>
    apiPatch<RecurringTransaction>(`/api/v1/personal/recurring/${id}`, data),

  triggerRecurring: (id: string) =>
    apiPatch<any>(`/api/v1/personal/recurring/${id}/trigger`),

  deactivateRecurring: (id: string) =>
    apiDelete<any>(`/api/v1/personal/recurring/${id}`),
};
